src/components/canvas-tools/drawing-tool/drawing-layer.tsx

```diff
@@ -0,0 +1,973 @@
+}
+
+interface DrawingLayerViewProps {
+  model: ToolTileModelType;
+  readOnly: boolean;
+}
+
+interface DrawingLayerViewState {
+  currentDrawingObject: LineObject|RectangleObject|EllipseObject|VectorObject|null;
+  objects: ObjectMap;
+  selectedObjects: DrawingObject[];
+  selectionBox: SelectionBox|null;
+  hoverObject: DrawingObject|null;
+  actionsCount: number;
+}
+
+export class DrawingLayerView extends React.Component<DrawingLayerViewProps, DrawingLayerViewState> {
+  public objects: ObjectMap;
+  public currentTool: DrawingTool|null;
+  public tools: DrawingToolMap;
+  private svgRef: React.RefObject<{}>|null;
+  private setSvgRef: (element: any) => void;
+
+  constructor(props: DrawingLayerViewProps){
+    super(props);
+
+    this.state = {
+      currentDrawingObject: null,
+      objects: {},
+      selectionBox: null,
+      selectedObjects: [],
+      hoverObject: null,
+      actionsCount: 0
+    };
+
+    this.tools = {
+      line: new LineDrawingTool(this),
+      vector: new VectorDrawingTool(this),
+      selection: new SelectionDrawingTool(this),
+      rectangle: new RectangleDrawingTool(this),
+      ellipse: new EllipseDrawingTool(this)
+    };
+    this.currentTool = this.tools.selection;
+
+    this.objects = {};
+
+    this.svgRef = null;
+    this.setSvgRef = (element) => {
+      this.svgRef = element;
+    };
+
+    this.addListeners();
+  }
+
+  public componentDidMount() {
+    this.syncChanges();
+  }
+
+  public componentDidUpdate(prevProps: DrawingLayerViewProps) {
+    const drawingContent = this.props.model.content as DrawingContentModelType;
+    const prevDrawingContent = prevProps.model.content as DrawingContentModelType;
+
+    this.syncChanges();
+
+    const newSettings = this.toolbarSettings(drawingContent);
+    const prevSettings = this.toolbarSettings(prevDrawingContent);
+    if (JSON.stringify(newSettings) !== JSON.stringify(prevSettings)) {
+      this.setCurrentToolSettings(newSettings);
+    }
+
+    if (drawingContent.selectedButton !== prevDrawingContent.selectedButton) {
+      switch (drawingContent.selectedButton) {
+        case "select":
+          this.setCurrentTool(this.tools.selection);
+          break;
+        case "line":
+          this.setCurrentTool((this.tools.line as LineDrawingTool).setSettings(newSettings));
+          break;
+        case "vector":
+          this.setCurrentTool((this.tools.vector as VectorDrawingTool).setSettings(newSettings));
+          break;
+        case "rectangle":
+          this.setCurrentTool((this.tools.rectangle as RectangleDrawingTool).setSettings(newSettings));
+          break;
+        case "ellipse":
+          this.setCurrentTool((this.tools.ellipse as EllipseDrawingTool).setSettings(newSettings));
+          break;
+      }
+    }
+  }
+
+  public addListeners() {
+    window.addEventListener("keyup", (e) => {
+      if (!this.props.readOnly) {
+        switch (e.keyCode) {
+          case 8:
+          case 46:
```

Comment:
Comment or constant would be useful here.

Reply:
:+1:

Reply:
I hate to bring this up, but as much as I like the switch to `e.key`, its behavior is not completely standardized across browsers yet. In particular, according to a Google Doc linked from [this discussion](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/8860571/) (which I found by following sources from [caniuse](https://www.caniuse.com/#search=keyboardevent.key)) Edge gets it wrong for the delete key. You could add "Del" to your switch statement, but safest might be to stick with the deprecated-but-functional `keyCode` along with a comment or locally-defined constants. I've been looking at these APIs a lot recently in implementing keyboard shortcuts for cut/copy/paste and it's mind-boggling that the options are deprecated APIs that work reliably and standardized ones that don't.